cmd/update-swt/main.go

```diff
@@ -0,0 +1,460 @@
+// https://download.eclipse.org/eclipse/downloads/, choose your version and
+// click on it. The version to pass to this script is the last URL segment and
+// should take the form "[RS]-<version>-<datetime>".
+// This tool uses a bunch of regex's to parse HTML and the .bzl files, and does
+// so in a dumb, but defensive manner. This means that trivial changes to the
+// Eclipse HTML or the .bzl files will likely make the parsing fail, but at
+// least if it succeeds, it's fairly confident that what it got is what it
+// thinks it got.
+package main
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/google/gapid/core/app"
+	"github.com/google/gapid/core/log"
+)
+
+var (
+	editSWT   = flag.Bool("swt", true, "Edit swt.bzl")
+	editJFace = flag.Bool("jface", true, "Edit jface.bzl")
+)
+
+const (
+	labelLinux   = "Linux (64 bit version)"
+	labelWindows = "Windows (64 bit version)"
+	labelMacOS   = "Mac OSX (64 bit version)"
+	swtBzl       = "tools/build/third_party/swt.bzl"
+	jfaceBzl     = "tools/build/third_party/jface.bzl"
+)
+
+var (
+	repoBase, _    = url.Parse("https://download.eclipse.org/eclipse/downloads/drops4/")
+	repoTableRegex = h3TableRegex("Repository")
+	swtTableRegex  = h3TableRegex("SWT")
+	tableRowRegex  = regexp.MustCompile("(?s:<tr(?: [^>]*)?>(.*?)</tr>)")
+	tableCellRegex = regexp.MustCompile("(?s:<t[dh](?: [^>]*)?>(.*?)</t[dh]>)")
+	linkRegex1     = regexp.MustCompile("<a(?: [^>]*)? href=\"([^\"]+)\"")
+	linkRegex2     = regexp.MustCompile("<a href='([^']+)'>(.*?)</a>")
+	pluginsRegex   = divRegex("dirlist")
+	swtBzlRegex    = regexp.MustCompile("(?m:^_(URL|SHA)_(LINUX|WIN|OSX) = \"([^\"]*)\"$)")
+	jfaceBaseRegex = regexp.MustCompile("(?m:^_BASE = \"[^\"]*\"$)")
+	jfaceBzlRegex  = regexp.MustCompile(
+		"(?s:struct\\(\\s*name = \"([^\"]+)\"," +
+			"\\s*version = \"([^\"]+)\"," +
+			"\\s*sha = \"([^\"]+)\"," +
+			"\\s*sha_src = \"([^\"]+)\"," +
+			")")
+)
+
+func main() {
+	app.ShortHelp = "update-swt updates the SWT and JFace repository rules"
+	app.ShortUsage = "<version>"
+	app.Name = "update-swt"
+	app.Run(run)
+}
+
+func run(ctx context.Context) error {
+	if flag.NArg() != 1 {
+		app.Usage(ctx, "Exactly one version expected, got %d", flag.NArg())
+		return nil
+	}
+	version := flag.Args()[0]
+
+	repo, err := getRepo(ctx, version)
+	if err != nil {
+		return log.Errf(ctx, err, "Failed to get repo for version '%s'", version)
+	}
+
+	if *editSWT {
+		if err := repo.editSWTBazel(ctx); err != nil {
+			return log.Errf(ctx, err, "Failed to edit the swt.bzl file")
+		}
+	}
+
+	if *editJFace {
+		if err := repo.editJFaceBazel(ctx); err != nil {
+			return log.Errf(ctx, err, "Failed to edit the jface.bzl file")
+		}
+	}
+
+	return nil
+}
+
+func imDumb(ctx context.Context, cause error, fmt string, args ...interface{}) error {
+	log.E(ctx, "I'm a really dumb script and something may have changed. I give up.")
```

Comment:
This logging function is used quite a lot lol ;) Just a little worry whether the following exceptions would happen at a certain high probability and get exposed to the users.

Reply:
Yeah, can you please add a slightly more helpful message here? Such as, please contact the AGI team, etc.